### PR TITLE
Fix L2 dst address computation (very intrusive)

### PR DESCRIPTION
**Pull request by gpotter2**
Two absolutely horrifying issues, that have been there since the beginning of times.

- **Currently if you do**:
```python
>>> pkt = Ether()/ARP(pdst="10.0.0.1")
>>> srp(pkt)
```
it sends a `ARP` request implicitly to resolve the Ethernet destination of the L2 dst... then sends it a who-has request. This makes no sense, that's two ARP who-has requests.

- **Currently if you do**:
```python
>>> pkt = Ether() / IP(pdst="10.0.0.1")
>>> pkt.show()
```
it also sends a `ARP` request implicitly. This looks like an unintended behavior, I'd argue the L2 dst should only be resolved  when building the packet (when using send, show2, etc.)

**This PR does two things:**
- in general, only resolve the Ethernet dst address when building the packet, and not just when printing it
- fix ARP which has a completly broken default behavior

### CODE REVIEW

File: scapy/layers/l2.py [hunk 1]
```diff
@@ -176,12 +182,10 @@ def i2h(self, pkt, x):
...
                 else:
                     x = "ff:ff:ff:ff:ff:ff"
-                    warning("Mac address to reach destination not found. Using broadcast.")  # noqa: E501
-        return super(DestMACField, self).i2h(pkt, x)
-
-    def i2m(self, pkt, x):
-        # type: (Optional[Packet], Optional[str]) -> bytes
-        return super(DestMACField, self).i2m(pkt, self.i2h(pkt, x))
+                    warning(
+                        "Mac address to reach destination not found. Using broadcast."
```
Comment (guedou) [inline scapy/layers/l2.py:186]:
```suggestion
                        "MAC address to reach destination not found. Using broadcast."
```

File: scapy/layers/l2.py [hunk 2]
```diff
@@ -311,8 +315,10 @@ class LLC(Packet):
                    ByteField("ctrl", 0)]
 
 
-def l2_register_l3(l2, l3):
-    # type: (Packet, Packet) -> Optional[str]
+def l2_register_l3(l2: Packet, l3: Packet) -> Optional[str]:
+    """
+    Delegates resolving the default L2 dst address to the payload of L3.
```
Comment (guedou) [inline scapy/layers/l2.py:320]:
```suggestion
    Delegates resolving the default L2 destination address to the payload of L3.
```

File: scapy/layers/l2.py [hunk 3]
```diff
@@ -554,15 +560,27 @@ def mysummary(self):
         return self.sprintf("ARP %op% %psrc% > %pdst%")
 
 
-def l2_register_l3_arp(l2, l3):
-    # type: (Packet, Packet) -> Optional[str]
-    # TODO: support IPv6?
-    plen = l3.plen if l3.plen is not None else l3.get_field("pdst").i2len(l3, l3.pdst)
+def l2_register_l3_arp(l2: Packet, l3: Packet) -> Optional[str]:
+    """
+    Resolves the default L2 dst address when ARP is used.
```
Comment (guedou) [inline scapy/layers/l2.py:565]:
```suggestion
    Resolves the default L2 destination address when ARP is used.
```

File: scapy/layers/l2.py [hunk 4]
```diff
@@ -554,15 +560,27 @@ def mysummary(self):
...
+    """
+    if l3.op == 1:  # who-has
+        return "ff:ff:ff:ff:ff:ff"
+    elif l3.op == 2:  # is-at
+        log_runtime.warning(
+            "You should be providing the Ethernet dst mac when sending is-at ARP."
```
Comment (guedou) [inline scapy/layers/l2.py:571]:
```suggestion
            "You should be providing the Ethernet destination MAC address when sending an is-at ARP."
```

File: scapy/layers/l2.py [hunk 5]
```diff
@@ -554,15 +560,27 @@ def mysummary(self):
...
     if plen == 4:
         return getmacbyip(l3.pdst)
+    elif plen == 32:
+        from scapy.layers.inet6 import getmacbyip6
+        return getmacbyip6(l3.pdst)
+    # Can't even do that
     log_runtime.warning(
-        "Unable to guess L2 MAC address from an ARP packet with a "
-        "non-IPv4 pdst. Provide it manually !"
+        "You should be providing the Ethernet dst mac when sending this "
```
Comment (guedou) [inline scapy/layers/l2.py:582]:
```suggestion
        "You should be providing the Ethernet destination MAC address when sending this "
```

File: scapy/layers/inet.py [hunk 6]
```diff
@@ -1099,6 +1099,9 @@ def mysummary(self):
 
 
 def inet_register_l3(l2, l3):
+    """
+    Resolves the default L2 dst address when IP is used.
```
Comment (guedou) [inline scapy/layers/inet.py:1103]:
```suggestion
    Resolves the default L2 destination address when IPv4 is used.
```

File: scapy/layers/inet6.py [hunk 7]
```diff
@@ -475,6 +475,9 @@ def dispatch_hook(cls, _pkt=None, *_, **kargs):
 
 
 def inet6_register_l3(l2, l3):
+    """
+    Resolves the default L2 dst address when IPv6 is used.
```
Comment (guedou) [inline scapy/layers/inet6.py:479]:
```suggestion
    Resolves the default L2 destination address when IPv6 is used.
```

**Review by guedou — DISMISSED** [review]
Good idea! This looks good to me.

This PR might break some scenarios where the packet was built then send later. If the host goes offline between the two, the Ether destination will be send to the broadcast address.

**Comment by gpotter2** [comment]
Thanks for the review, I've updated the PR.

> If the host goes offline between the two, the Ether destination will be send to the broadcast address.

I would argue that this is not an issue.

**Review by guedou — APPROVED** [review]

**Pull request merged**